docs/architecture/ErrorReporting.md

```diff
@@ -0,0 +1,140 @@
+# Handling and reporting errors
+
+    WARNING: This document describes future plans for error handling.
+    The UI library does not (yet) implement error handling as described here.
+
+This document describes how errors are handled within the ACS UI library and how they are reported to the surrounding application.
+
+## Error sources
+
+There are two categories of errors, and they are handled differently.
+
+1. ACS backend errors: Errors returned by the REST calls to ACS backend services.
+
+   1. These can be transient or permanent (special case: permission error).
+
+2. UI errors: Programming errors in the UI library / errors returned from other browser APIs
+
+## ACS backend errors
+
+Handling ACS backend errors is challenging when using either the composites or the pure components with the react bindings provided by the ACS UI library:
+
+* ACS backend API calls are asynchronous
+* ACS backend API calls are often initiated outside of the React component lifecycle methods (e.g on interaction handler callbacks or `useEffect` blocks).
+
+But in both cases, all ACS backend API calls are mediated by the (chat and calling) stateful clients. Thus, the stateful client tees errors to a special field in the state:
+
+```ts
+interface CallClientState {
+    // ... other fields ...
+    errors: { [key: SdkClientMethod]: ErrorState[];
+}
+
+type SdkClientMethod = 'CallClient.createCallAgent' | 'CallClient.getDeviceManager'; // ... and many more
+
+interface ErrorState {
+    /**
+     * Locally generated unique ID for error.
+     * Allows for de-duplicating similar errors from repeated API calls.
+     */
+    id: string;
+    error: Error;
+}
+```
+
+Stateful client API method calls continue to return errors (e.g. as a failed promise). So they can be handled directly as well.
+
+The stateful clients stores the most recent errors per method, evicting older errors if necessary. The number of errors stored can be configured, or set to -1 to disable deletion of older errors. e.g., the composites disable error eviction in the stateful client and delete older errors themselves. By default, only the most recent error is stored.
```

Comment:
What's an example of when a user uses the last n errors stored in our state instead of just using the onError event?

Reply:
When using pure components, there is no `onError` event.

Similarly, the `ErrorBar` component will directly work on the stateful `errors` field and might want to show more than one error.

Reply:
Pure components shouldn't give errors though (except for category 2 unexpected errors - but they're category 2 and "aren't handled via the mechanism used for ACS backend errors." and should use the react error boundary)

So does this state only exist to pipe errors into the ErrorBar component for composites? What if instead of an errorbar relying on state - we have specific state for error bar errors, as I'm sure most errors shouldn't be presented to an end user - only ones they can act upon

Reply:
(My aim here is to avoid us having to expose and maintain an extra error interface in the state - that sounds difficult to use as errors a developer/user might care about might suddenly disappear because new ones evict them)

Reply:
By pure components I mean pure components + bindings (i.e., building without the adapters).
And we're still discussing category 1 errors here.

Two explicit use-cases so far:
- `ErrorBar`
- Disable `SendBox` when error indicates we can't send messages (e.g. when backend throttles us).

Think of this as doing for errors exactly what we're doing for state -- converting the response from API calls to data in a way that reactive UI can bind to the data.

As for error eviction, as stated, it is possible to configure the stateful client to never evict any errors and fully control when errors are dropped. But more likely than not, the default of keeping the last 1 error will suffice for most users, especially because this will save the last error per method.

Reply:
Ah I was reading all this through the lens of composite-only. I really like having last n _per method_ - that makes a lot of sense and much easier to use (think I missed that in my last read through and it was just last n errors of any kind)